smartcontract/service/native/param.go

```diff
@@ -98,29 +105,58 @@ func SetParam(native *NativeService) error {
 		return errors.NewErr("[Set Param]Deserialize failed!")
 	}
 	for _, param := range params.ParamList {
-		deleteParamInCache(param.K) // delete the param in cache, in case of data inconsistencies
-		native.CloneCache.Add(scommon.ST_STORAGE, getParamKey(contract, param.K), getParamStorageItem(param.V))
-		notifyParamSetSucess(native, contract, param)
+		native.CloneCache.Add(scommon.ST_STORAGE, getParamKey(contract, param.K, PREPARE_VALUE),
+			getParamStorageItem(param.V))
+		notifyParamSetSuccess(native, contract, param)
+	}
+	return nil
+}
+
+func EnforceParam(native *NativeService) error {
+	contract := native.ContextRef.CurrentContext().ContractAddress
+	initAdmin(native, contract)
+	if !native.ContextRef.CheckWitness(admin.Address) {
+		return errors.NewErr("[Enforce Param]Authentication failed!")
+	}
+	params := new(states.Params)
+	if err := params.Deserialize(bytes.NewBuffer(native.Input)); err != nil {
+		return errors.NewErr("[Enforce Param]Deserialize failed!")
+	}
+	for _, param := range params.ParamList {
+		paramName := param.K
+		// read prepare value
+		value, err := native.CloneCache.Get(scommon.ST_STORAGE,
+			getParamKey(genesis.ParamContractAddress, paramName, PREPARE_VALUE))
+		if err != nil {
+			return errors.NewDetailErr(err, errors.ErrNoCode, "[Enforce Param] storage error!")
+		}
+		if value == nil {
+			return errors.NewErr("[Enforce Param] Prepare param doesn't exist!")
+		}
+		item, ok := value.(*cstates.StorageItem)
+		if !ok {
+			return errors.NewDetailErr(err, errors.ErrNoCode, "[Enforce Param] storage error!")
+		}
+		// set prepare value to current value, make it effective
+		native.CloneCache.Add(scommon.ST_STORAGE, getParamKey(contract, paramName, CURRENT_VALUE), value)
+		paramValue := string(item.Value)
+		setParamToCache(paramName, paramValue)
```

Comment:
if the transaction execute failed in this for loop. this may have problem, because you have modified some of the key in memory, but all of them will keep unchanged in db.